src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,1209 @@
+                                            exceptions.Add(e);
+                                            skippedNodes[database].Add(selectedNode.ClusterTag);
+                                        }
+
+                                        if (newWorker != null)
+                                        {
+                                            break;
+                                        }
+                                    }
+                                }
+                            }
+
+                            if (newWorker == null)
+                            {
+                                // we could not assert connection to worker, have no nodes left to connect and could not create new worker => setup connect with timeout and continue to next worker.
+                                _rejectedShards.Add(database);
+                            }
+                            else
+                            {
+                                // we need to assert the connection again
+                                workers.Add(newWorker);
+                            }
+                        }
+                        else
+                        {
+                            throw new Exception($"TODO: egor, assert connection status failed in InitAsync");
+                        }
+                    }
+                }
+            }
+
+            if (sp.ElapsedMilliseconds > timeout)
+                throw new Exception("TODO: egor got timeout in assert connection status");
+        }
+
+        private async Task SendOptionToServerWorker(SubscriptionServerWorker worker)
+        {
+            using (_serverStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                using (var optionsJson = DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(Options, context))
+                {
+                    await optionsJson.WriteJsonToAsync(worker.Stream, CancellationTokenSource.Token).ConfigureAwait(false);
+                    await worker.Stream.FlushAsync(CancellationTokenSource.Token).ConfigureAwait(false);
+                }
+            }
+        }
+
+        private async Task ProcessSubscriptionAsync()
+        {
+            var replyFromClientTask = GetReplyFromClientAsync(TcpConnection, _copiedBuffer.Buffer, _isDisposed);
+            var i = 0;
+            using (DisposeOnDisconnect)
+            {
+                while (CancellationTokenSource.IsCancellationRequested == false)
+                {
+                    var workers = _workers;
+                    if (workers.Count == 0)
+                    {
+                        await Task.Delay(1000);
+                        continue;
+                    }
+
+                    var currentWorker = workers[i++ % workers.Count];
+                    currentWorker.PullingTask ??= ReadNextObject(currentWorker.Context, currentWorker, currentWorker.Buffer);
```

Comment:
You are pulling from the next client while waiting for the reply from the client? 
Why not wait until we actually get the reply? Where is the data sitting in the meantime? If the batch size is high, the TCP buffers would be full and drop, no?